// @permitio/cli/dist/lib/api.js
import { PERMIT_API_URL, PERMIT_ORIGIN_URL } from '../config.js';
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const apiCall = async (endpoint, token, cookie, method = 'GET', body) => {
    let defaultResponse = {
        headers: new Headers(),
        response: {},
        status: -1,
        error: null,
    };
    const options = {
        method,
        headers: {
            Accept: '*/*',
            Origin: PERMIT_ORIGIN_URL,
            Authorization: `Bearer ${token}`,
            Cookie: cookie ?? '',
            'Content-Type': 'application/json',
        },
    };
    if (body) {
        options.body = body;
    }
    try {
        const res = await fetch(`${PERMIT_API_URL}/${endpoint}`, options);
        if (!res.ok) {
            const errorText = await res.json();
            defaultResponse.error = `Request failed with status ${res.status}: ${errorText.message ?? errorText.detail}`;
            defaultResponse.status = res.status;
        }
        else {
            const response = await res.json();
            defaultResponse.headers = res.headers;
            defaultResponse.response = response;
            defaultResponse.status = res.status;
        }
    }
    catch (error) {
        defaultResponse.error =
            error instanceof Error ? error.message : 'Unknown fetch error occurred';
    }
    return defaultResponse;
};
