// permitio/build/main/api/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseFactsPermitAPI = exports.BasePermitApi = exports.PermitApiError = void 0;
const axios_1 = __importDefault(require("axios"));
const openapi_1 = require("../openapi");
const base_1 = require("../openapi/base");
const context_1 = require("./context");
class PermitApiError extends Error {
    constructor(message, originalError) {
        super(message);
        this.originalError = originalError;
    }
    get formattedAxiosError() {
        var _a;
        return {
            code: this.originalError.code,
            message: this.message,
            error: (_a = this.originalError.response) === null || _a === void 0 ? void 0 : _a.data,
            status: this.originalError.status,
        };
    }
    get request() {
        return this.originalError.request;
    }
    get response() {
        return this.originalError.response;
    }
}
exports.PermitApiError = PermitApiError;
class BasePermitApi {
    constructor(config, logger) {
        var _a;
        this.config = config;
        this.logger = logger;
        const version = (_a = process.env.npm_package_version) !== null && _a !== void 0 ? _a : 'unknown';
        this.openapiClientConfig = new openapi_1.Configuration({
            basePath: `${this.config.apiUrl}`,
            accessToken: this.config.token,
            baseOptions: {
                headers: {
                    'X-Permit-SDK-Version': `node:${version}`,
                },
            },
        });
        this.scopeApi = new openapi_1.APIKeysApi(this.openapiClientConfig, base_1.BASE_PATH, this.config.axiosInstance);
    }
    /**
     * Sets the API context and permitted access level based on the API key scope.
     */
    async setContextFromApiKey() {
        var _a, _b;
        try {
            this.logger.debug('Fetching api key scope');
            const response = await this.scopeApi.getApiKeyScope();
            if (response.data.organization_id !== undefined && response.data.organization_id !== null) {
                this.config.apiContext._saveApiKeyAccessibleScope(response.data.organization_id, response.data.project_id, response.data.environment_id);
                if (response.data.project_id !== undefined && response.data.project_id !== null) {
                    if (response.data.environment_id !== undefined && response.data.environment_id !== null) {
                        // set environment level context
                        this.logger.debug(`setting: environment-level api context`);
                        this.config.apiContext.setEnvironmentLevelContext(response.data.organization_id, response.data.project_id, response.data.environment_id);
                        return;
                    }
                    // set project level context
                    this.logger.debug(`setting: project-level api context`);
                    this.config.apiContext.setProjectLevelContext(response.data.organization_id, response.data.project_id);
                    return;
                }
                // set org level context
                this.logger.debug(`setting: organization-level api context`);
                this.config.apiContext.setOrganizationLevelContext(response.data.organization_id);
                return;
            }
            throw new context_1.PermitContextError('could not set api context level');
        }
        catch (err) {
            if (axios_1.default.isAxiosError(err)) {
                this.logger.error(`[${(_a = err === null || err === void 0 ? void 0 : err.response) === null || _a === void 0 ? void 0 : _a.status}] permit.api.getApiKeyScope(), err: ${JSON.stringify((_b = err === null || err === void 0 ? void 0 : err.response) === null || _b === void 0 ? void 0 : _b.data)}`);
            }
            throw new context_1.PermitContextError('could not fetch the api key scope in order to set the api context level');
        }
    }
    /**
     * Ensure that the API Key has the necessary permissions to successfully call the API endpoint.
     * Note that this check is not foolproof, and the API may still throw 401.
     * @param requiredAccessLevel The required API Key Access level for the endpoint.
     * @throws PermitContextError If the currently set API key access level does not match the required access level.
     */
    async ensureAccessLevel(requiredAccessLevel) {
        // should only happen once in the lifetime of the SDK
        if (this.config.apiContext.contextLevel === context_1.ApiContextLevel.WAIT_FOR_INIT ||
            this.config.apiContext.permittedAccessLevel === context_1.ApiKeyLevel.WAIT_FOR_INIT) {
            await this.setContextFromApiKey();
        }
        if (requiredAccessLevel !== this.config.apiContext.permittedAccessLevel) {
            if (context_1.API_ACCESS_LEVELS.indexOf(requiredAccessLevel) <
                context_1.API_ACCESS_LEVELS.indexOf(this.config.apiContext.permittedAccessLevel)) {
                throw new context_1.PermitContextError(`You're trying to use an SDK method that requires an API Key with access level: ${requiredAccessLevel}, ` +
                    `however the SDK is running with an API key with level ${this.config.apiContext.permittedAccessLevel}.`);
            }
        }
    }
    /**
     * Ensure that the API context matches the required endpoint context.
     * @param requiredContext The required API context level for the endpoint.
     * @throws PermitContextError If the currently set API context level does not match the required context level.
     */
    async ensureContext(requiredContext) {
        // should only happen once in the lifetime of the SDK
        if (this.config.apiContext.contextLevel === context_1.ApiContextLevel.WAIT_FOR_INIT ||
            this.config.apiContext.permittedAccessLevel === context_1.ApiKeyLevel.WAIT_FOR_INIT) {
            await this.setContextFromApiKey();
        }
        if (this.config.apiContext.contextLevel < requiredContext ||
            this.config.apiContext.contextLevel === context_1.ApiContextLevel.WAIT_FOR_INIT) {
            throw new context_1.PermitContextError(`You're trying to use an SDK method that requires an API context of ${context_1.ApiContextLevel[requiredContext]}, ` +
                `however the SDK is running in a less specific context level: ${context_1.ApiContextLevel[this.config.apiContext.contextLevel]}.`);
        }
    }
    handleApiError(err) {
        var _a, _b, _c;
        if (axios_1.default.isAxiosError(err)) {
            // this is an http response with an error status code
            const logMessage = `Got error status code: ${(_a = err.response) === null || _a === void 0 ? void 0 : _a.status}, err: ${JSON.stringify((_b = err === null || err === void 0 ? void 0 : err.response) === null || _b === void 0 ? void 0 : _b.data)}`;
            const apiMessage = (_c = err.response) === null || _c === void 0 ? void 0 : _c.data.message;
            // log this to the SDK logger
            this.logger.error(logMessage);
            // and throw a permit error exception
            throw new PermitApiError(apiMessage, err);
        }
        else {
            // unexpected error, just throw
            throw err;
        }
    }
}
exports.BasePermitApi = BasePermitApi;
class BaseFactsPermitAPI extends BasePermitApi {
    constructor(config, logger) {
        super(config, logger);
        this.config = config;
        this.logger = logger;
        if (config.proxyFactsViaPdp) {
            this.openapiClientConfig = new openapi_1.Configuration({
                basePath: `${this.config.pdp}`,
                accessToken: this.config.token,
                baseOptions: {
                    headers: Object.assign(Object.assign(Object.assign({}, this.openapiClientConfig.baseOptions.headers), (this.config.factsSyncTimeout !== null && {
                        'X-Wait-Timeout': this.config.factsSyncTimeout.toString(),
                    })), (this.config.factsSyncTimeoutPolicy && {
                        'X-Timeout-Policy': this.config.factsSyncTimeoutPolicy,
                    })),
                },
            });
        }
    }
    clone() {
        return new this.constructor(this.config, this.logger);
    }
    waitForSync(timeout, policy) {
        if (this.config.proxyFactsViaPdp) {
            const clone = this.clone();
            clone.openapiClientConfig.baseOptions.headers['X-Wait-Timeout'] =
                timeout === null ? '' : timeout.toString();
            const timeoutPolicy = policy || this.config.factsSyncTimeoutPolicy;
            if (timeoutPolicy) {
                clone.openapiClientConfig.baseOptions.headers['X-Timeout-Policy'] = timeoutPolicy;
            }
            return clone;
        }
        else {
            this.logger.warn("Attempted to wait for sync, but 'proxyFactsViaPdp' is not enabled. Ignoring.");
            return this;
        }
    }
}
exports.BaseFactsPermitAPI = BaseFactsPermitAPI;
