// permitio/build/module/api/relationship-tuples.js
import { RelationshipTuplesApi as AutogenRelationshipTuplesApi, } from '../openapi';
import { BASE_PATH } from '../openapi/base';
import { BaseFactsPermitAPI } from './base';
import { ApiContextLevel, ApiKeyLevel } from './context';
/**
 * The RelationshipTuplesApi class provides methods for interacting with Role createments.
 */
export class RelationshipTuplesApi extends BaseFactsPermitAPI {
    relationshipTuples;
    /**
     * Creates an instance of the RelationshipTuplesApi.
     * @param config - The configuration object for the Permit SDK.
     * @param logger - The logger instance for logging.
     */
    constructor(config, logger) {
        super(config, logger);
        this.relationshipTuples = new AutogenRelationshipTuplesApi(this.openapiClientConfig, BASE_PATH, this.config.axiosInstance);
    }
    /**
     * Retrieves a list of role createments based on the specified filters.
     *
     * @param params - The filters and pagination options for listing role createments.
     * @returns A promise that resolves with an array of role createments.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async list(params) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            return (await this.relationshipTuples.listRelationshipTuples({
                ...params,
                ...this.config.apiContext.environmentContext,
            })).data;
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
    /**
     * Creates a new relationship tuple, that states that a relationship (of type: relation)
     * exists between two resource instances: the subject and the object.
     *
     * @param tuple - The tuple to create
     * @returns A promise that resolves to the created relationship tuple.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async create(tuple) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            return (await this.relationshipTuples.createRelationshipTuple({
                ...this.config.apiContext.environmentContext,
                relationshipTupleCreate: tuple,
            })).data;
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
    /**
     * Removes a relationship tuple.
     *
     * @param tuple - The tuple to delete
     * @returns A promise that resolves when the tuple is successfully deleted.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async delete(tuple) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            return (await this.relationshipTuples.deleteRelationshipTuple({
                ...this.config.apiContext.environmentContext,
                relationshipTupleDelete: tuple,
            })).data;
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
    /**
     * Creates a new relationship tuple, that states that a relationship (of type: relation)
     * exists between two resource instances: the subject and the object.
     *
     * @returns A promise that resolves to the created relationship tuple.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     * @param tuples
     */
    async bulkRelationshipTuples(tuples) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            return (await this.relationshipTuples.bulkCreateRelationshipTuples({
                ...this.config.apiContext.environmentContext,
                relationshipTupleCreateBulkOperation: {
                    operations: tuples,
                },
            })).data;
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
    /**
     * Deletes multiple relationship tuples at once using the provided tuple data.
     * Each tuple object is of type RelationshipTupleDelete and is essentially a tuple of (subject, relation, object).
     *
     * @param tuples - he relationship tuples to delete.
     * @returns A promise that resolves with the bulk un relationship tuples report.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async bulkUnRelationshipTuples(tuples) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            return (await this.relationshipTuples.bulkDeleteRelationshipTuples({
                ...this.config.apiContext.environmentContext,
                relationshipTupleDeleteBulkOperation: {
                    idents: tuples,
                },
            })).data;
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
}
