// permitio/build/module/api/tenants.mjs
import { TenantsApi as AutogenTenantsApi, } from '../openapi';
import { BASE_PATH } from '../openapi/base';
import { BaseFactsPermitAPI } from './base';
import { ApiContextLevel, ApiKeyLevel } from './context';
/**
 * The TenantsApi class provides methods for interacting with Permit Tenants.
 */
export class TenantsApi extends BaseFactsPermitAPI {
    tenants;
    /**
     * Creates an instance of the TenantsApi.
     * @param config - The configuration object for the Permit SDK.
     * @param logger - The logger instance for logging.
     */
    constructor(config, logger) {
        super(config, logger);
        this.tenants = new AutogenTenantsApi(this.openapiClientConfig, BASE_PATH, this.config.axiosInstance);
    }
    /**
     * Retrieves a list of tenants.
     *
     * @param params Filtering and pagination options, @see {@link IListTenantsParams}
     * @returns A promise that resolves to an array of tenants.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async list(params) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            return (await this.tenants.listTenants({
                ...params,
                ...this.config.apiContext.environmentContext,
            })).data;
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
    /**
     * Retrieves a list of users for a given tenant.
     *
     * @param tenantKey - The key of the tenant for which to list users.
     * @param params - pagination and filtering params.
     * @returns A promise that resolves to a PaginatedResultUserRead object containing the list of tenant users.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async listTenantUsers({ tenantKey, ...params }) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            return (await this.tenants.listTenantUsers({
                ...params,
                ...this.config.apiContext.environmentContext,
                tenantId: tenantKey,
            })).data;
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
    /**
     * Retrieves a tenant by its key.
     *
     * @param tenantKey The key of the tenant.
     * @returns A promise that resolves to the tenant.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async get(tenantKey) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            return (await this.tenants.getTenant({
                ...this.config.apiContext.environmentContext,
                tenantId: tenantKey,
            })).data;
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
    /**
     * Retrieves a tenant by its key.
     * Alias for the {@link get} method.
     *
     * @param tenantKey The key of the tenant.
     * @returns A promise that resolves to the tenant.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async getByKey(tenantKey) {
        return await this.get(tenantKey);
    }
    /**
     * Retrieves a tenant by its ID.
     * Alias for the {@link get} method.
     *
     * @param tenantId The ID of the tenant.
     * @returns A promise that resolves to the tenant.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async getById(tenantId) {
        return await this.get(tenantId);
    }
    /**
     * Creates a new tenant.
     *
     * @param tenantData The data for the new tenant.
     * @returns A promise that resolves to the created tenant.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async create(tenantData) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            return (await this.tenants.createTenant({
                ...this.config.apiContext.environmentContext,
                tenantCreate: tenantData,
            })).data;
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
    /**
     * Updates a tenant.
     *
     * @param tenantKey The key of the tenant.
     * @param tenantData The updated data for the tenant.
     * @returns A promise that resolves to the updated tenant.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async update(tenantKey, tenantData) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            return (await this.tenants.updateTenant({
                ...this.config.apiContext.environmentContext,
                tenantId: tenantKey,
                tenantUpdate: tenantData,
            })).data;
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
    /**
     * Deletes a tenant.
     *
     * @param tenantKey The key of the tenant to delete.
     * @returns A promise that resolves when the tenant is deleted.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async delete(tenantKey) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            await this.tenants.deleteTenant({
                ...this.config.apiContext.environmentContext,
                tenantId: tenantKey,
            });
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
    /**
     * Deletes a user from a given tenant (also removes all roles granted to the user in that tenant).
     *
     * @param tenantKey - The key of the tenant from which the user will be deleted.
     * @param userKey - The key of the user to be deleted.
     * @returns A promise that resolves when the user is successfully deleted.
     * @throws {@link PermitApiError} If the API returns an error HTTP status code.
     * @throws {@link PermitContextError} If the configured {@link ApiContext} does not match the required endpoint context.
     */
    async deleteTenantUser(tenantKey, userKey) {
        await this.ensureAccessLevel(ApiKeyLevel.ENVIRONMENT_LEVEL_API_KEY);
        await this.ensureContext(ApiContextLevel.ENVIRONMENT);
        try {
            await this.tenants.deleteTenantUser({
                ...this.config.apiContext.environmentContext,
                tenantId: tenantKey,
                userId: userKey,
            });
        }
        catch (err) {
            this.handleApiError(err);
        }
    }
}
