// permitio/build/main/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigFactory = void 0;
const axios_1 = __importDefault(require("axios"));
const lodash_1 = __importDefault(require("lodash"));
const context_1 = require("./api/context");
/**
 * A factory class for the Permit SDK configuration
 */
class ConfigFactory {
    /**
     * @returns the default SDK configuration
     */
    static defaults() {
        return {
            token: lodash_1.default.get(process.env, 'PERMIT_API_KEY', ''),
            pdp: lodash_1.default.get(process.env, 'PERMIT_PDP_URL', 'http://localhost:7766'),
            apiUrl: lodash_1.default.get(process.env, 'PERMIT_API_URL', 'https://api.permit.io'),
            log: {
                // log level, default is warn (warnings and errors)
                level: lodash_1.default.get(process.env, 'PERMIT_LOG_LEVEL', 'warn'),
                // Label added to logs
                label: lodash_1.default.get(process.env, 'PERMIT_LOG_LABEL', 'Permit.io'),
                // When logging - dump full data to console as JSON
                json: JSON.parse(lodash_1.default.get(process.env, 'PERMIT_LOG_JSON', 'false')),
            },
            multiTenancy: {
                defaultTenant: 'default',
                useDefaultTenantIfEmpty: true,
            },
            timeout: undefined,
            throwOnError: true,
            apiContext: new context_1.ApiContext(),
            axiosInstance: axios_1.default.create(),
            proxyFactsViaPdp: false,
            factsSyncTimeout: null,
        };
    }
    /**
     * Builds the Permit SDK configuration from the values provided by the SDK user
     * and from the default SDK configuration when no specific values are set.
     *
     * @param options - a partial configuration
     * @returns the SDK configuration (for unset values returns the default config)
     */
    static build(options) {
        const config = lodash_1.default.merge(lodash_1.default.assign({}, ConfigFactory.defaults()), options);
        return config;
    }
}
exports.ConfigFactory = ConfigFactory;
