// permitio/build/module/openapi/api/resources-api.js
/* tslint:disable */
/* eslint-disable */
/**
 * Permit.io API
 *  Authorization as a service
 *
 * The version of the OpenAPI document: 2.0.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
import globalAxios from 'axios';
// Some imports not used depending on template conditions
// @ts-ignore
import { DUMMY_BASE_URL, assertParamExists, setBearerAuthToObject, setSearchParams, serializeDataIfNeeded, toPathString, createRequestFunction, } from '../common';
// @ts-ignore
import { BASE_PATH, BaseAPI } from '../base';
/**
 * ResourcesApi - axios parameter creator
 * @export
 */
export const ResourcesApiAxiosParamCreator = function (configuration) {
    return {
        /**
         * Creates a new resource (a type of object you may protect with permissions).
         * @summary Create Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {ResourceCreate} resourceCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createResource: async (projId, envId, resourceCreate, options = {}) => {
            // verify required parameter 'projId' is not null or undefined
            assertParamExists('createResource', 'projId', projId);
            // verify required parameter 'envId' is not null or undefined
            assertParamExists('createResource', 'envId', envId);
            // verify required parameter 'resourceCreate' is not null or undefined
            assertParamExists('createResource', 'resourceCreate', resourceCreate);
            const localVarPath = `/v2/schema/{proj_id}/{env_id}/resources`
                .replace(`{${'proj_id'}}`, encodeURIComponent(String(projId)))
                .replace(`{${'env_id'}}`, encodeURIComponent(String(envId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = { method: 'POST', ...baseOptions, ...options };
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication HTTPBearer required
            // http bearer authentication required
            await setBearerAuthToObject(localVarHeaderParameter, configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            setSearchParams(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = {
                ...localVarHeaderParameter,
                ...headersFromBaseOptions,
                ...options.headers,
            };
            localVarRequestOptions.data = serializeDataIfNeeded(resourceCreate, localVarRequestOptions, configuration);
            return {
                url: toPathString(localVarUrlObj),
                options: localVarRequestOptions,
            };
        },
        /**
         * Deletes the resource and all its related data.
         * @summary Delete Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteResource: async (projId, envId, resourceId, options = {}) => {
            // verify required parameter 'projId' is not null or undefined
            assertParamExists('deleteResource', 'projId', projId);
            // verify required parameter 'envId' is not null or undefined
            assertParamExists('deleteResource', 'envId', envId);
            // verify required parameter 'resourceId' is not null or undefined
            assertParamExists('deleteResource', 'resourceId', resourceId);
            const localVarPath = `/v2/schema/{proj_id}/{env_id}/resources/{resource_id}`
                .replace(`{${'proj_id'}}`, encodeURIComponent(String(projId)))
                .replace(`{${'env_id'}}`, encodeURIComponent(String(envId)))
                .replace(`{${'resource_id'}}`, encodeURIComponent(String(resourceId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = { method: 'DELETE', ...baseOptions, ...options };
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication HTTPBearer required
            // http bearer authentication required
            await setBearerAuthToObject(localVarHeaderParameter, configuration);
            setSearchParams(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = {
                ...localVarHeaderParameter,
                ...headersFromBaseOptions,
                ...options.headers,
            };
            return {
                url: toPathString(localVarUrlObj),
                options: localVarRequestOptions,
            };
        },
        /**
         * Gets a single resource, if such resource exists.
         * @summary Get Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getResource: async (projId, envId, resourceId, options = {}) => {
            // verify required parameter 'projId' is not null or undefined
            assertParamExists('getResource', 'projId', projId);
            // verify required parameter 'envId' is not null or undefined
            assertParamExists('getResource', 'envId', envId);
            // verify required parameter 'resourceId' is not null or undefined
            assertParamExists('getResource', 'resourceId', resourceId);
            const localVarPath = `/v2/schema/{proj_id}/{env_id}/resources/{resource_id}`
                .replace(`{${'proj_id'}}`, encodeURIComponent(String(projId)))
                .replace(`{${'env_id'}}`, encodeURIComponent(String(envId)))
                .replace(`{${'resource_id'}}`, encodeURIComponent(String(resourceId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = { method: 'GET', ...baseOptions, ...options };
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication HTTPBearer required
            // http bearer authentication required
            await setBearerAuthToObject(localVarHeaderParameter, configuration);
            setSearchParams(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = {
                ...localVarHeaderParameter,
                ...headersFromBaseOptions,
                ...options.headers,
            };
            return {
                url: toPathString(localVarUrlObj),
                options: localVarRequestOptions,
            };
        },
        /**
         * Lists all the resources defined in your schema.
         * @summary List Resources
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {boolean} [includeBuiltIn] Whether to include or exclude built-in resources, default is False
         * @param {number} [page] Page number of the results to fetch, starting at 1.
         * @param {number} [perPage] The number of results per page (max 100).
         * @param {includeTotalCount} [includeTotalCount] Include the total count of resources in the response, default is False
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        listResources: async (projId, envId, includeBuiltIn, page, perPage, includeTotalCount, options = {}) => {
            // verify required parameter 'projId' is not null or undefined
            assertParamExists('listResources', 'projId', projId);
            // verify required parameter 'envId' is not null or undefined
            assertParamExists('listResources', 'envId', envId);
            const localVarPath = `/v2/schema/{proj_id}/{env_id}/resources`
                .replace(`{${'proj_id'}}`, encodeURIComponent(String(projId)))
                .replace(`{${'env_id'}}`, encodeURIComponent(String(envId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = { method: 'GET', ...baseOptions, ...options };
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication HTTPBearer required
            // http bearer authentication required
            await setBearerAuthToObject(localVarHeaderParameter, configuration);
            if (includeBuiltIn !== undefined) {
                localVarQueryParameter['include_built_in'] = includeBuiltIn;
            }
            if (page !== undefined) {
                localVarQueryParameter['page'] = page;
            }
            if (perPage !== undefined) {
                localVarQueryParameter['per_page'] = perPage;
            }
            if (includeTotalCount !== undefined) {
                localVarQueryParameter['include_total_count'] = includeTotalCount;
            }
            setSearchParams(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = {
                ...localVarHeaderParameter,
                ...headersFromBaseOptions,
                ...options.headers,
            };
            return {
                url: toPathString(localVarUrlObj),
                options: localVarRequestOptions,
            };
        },
        /**
         * Completely replaces the resource definition.  - If the resource key is changed, all role and permissions assignments for the the resource will be revoked. - If the resource key is unchanged, but some actions are removed or renamed from the resource definition, role and permissions assignments for these actions will be revoked.  TODO: we need to decide if we are auto-revoking, or if we are rejecting the PUT completely while there are permissions that can be affected.
         * @summary Replace Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {ResourceReplace} resourceReplace
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        replaceResource: async (projId, envId, resourceId, resourceReplace, options = {}) => {
            // verify required parameter 'projId' is not null or undefined
            assertParamExists('replaceResource', 'projId', projId);
            // verify required parameter 'envId' is not null or undefined
            assertParamExists('replaceResource', 'envId', envId);
            // verify required parameter 'resourceId' is not null or undefined
            assertParamExists('replaceResource', 'resourceId', resourceId);
            // verify required parameter 'resourceReplace' is not null or undefined
            assertParamExists('replaceResource', 'resourceReplace', resourceReplace);
            const localVarPath = `/v2/schema/{proj_id}/{env_id}/resources/{resource_id}`
                .replace(`{${'proj_id'}}`, encodeURIComponent(String(projId)))
                .replace(`{${'env_id'}}`, encodeURIComponent(String(envId)))
                .replace(`{${'resource_id'}}`, encodeURIComponent(String(resourceId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = { method: 'PUT', ...baseOptions, ...options };
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication HTTPBearer required
            // http bearer authentication required
            await setBearerAuthToObject(localVarHeaderParameter, configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            setSearchParams(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = {
                ...localVarHeaderParameter,
                ...headersFromBaseOptions,
                ...options.headers,
            };
            localVarRequestOptions.data = serializeDataIfNeeded(resourceReplace, localVarRequestOptions, configuration);
            return {
                url: toPathString(localVarUrlObj),
                options: localVarRequestOptions,
            };
        },
        /**
         * Partially updates the resource definition. Fields that will be provided will be completely overwritten.
         * @summary Update Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {ResourceUpdate} resourceUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateResource: async (projId, envId, resourceId, resourceUpdate, options = {}) => {
            // verify required parameter 'projId' is not null or undefined
            assertParamExists('updateResource', 'projId', projId);
            // verify required parameter 'envId' is not null or undefined
            assertParamExists('updateResource', 'envId', envId);
            // verify required parameter 'resourceId' is not null or undefined
            assertParamExists('updateResource', 'resourceId', resourceId);
            // verify required parameter 'resourceUpdate' is not null or undefined
            assertParamExists('updateResource', 'resourceUpdate', resourceUpdate);
            const localVarPath = `/v2/schema/{proj_id}/{env_id}/resources/{resource_id}`
                .replace(`{${'proj_id'}}`, encodeURIComponent(String(projId)))
                .replace(`{${'env_id'}}`, encodeURIComponent(String(envId)))
                .replace(`{${'resource_id'}}`, encodeURIComponent(String(resourceId)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = { method: 'PATCH', ...baseOptions, ...options };
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication HTTPBearer required
            // http bearer authentication required
            await setBearerAuthToObject(localVarHeaderParameter, configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            setSearchParams(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = {
                ...localVarHeaderParameter,
                ...headersFromBaseOptions,
                ...options.headers,
            };
            localVarRequestOptions.data = serializeDataIfNeeded(resourceUpdate, localVarRequestOptions, configuration);
            return {
                url: toPathString(localVarUrlObj),
                options: localVarRequestOptions,
            };
        },
    };
};
/**
 * ResourcesApi - functional programming interface
 * @export
 */
export const ResourcesApiFp = function (configuration) {
    const localVarAxiosParamCreator = ResourcesApiAxiosParamCreator(configuration);
    return {
        /**
         * Creates a new resource (a type of object you may protect with permissions).
         * @summary Create Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {ResourceCreate} resourceCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        async createResource(projId, envId, resourceCreate, options) {
            const localVarAxiosArgs = await localVarAxiosParamCreator.createResource(projId, envId, resourceCreate, options);
            return createRequestFunction(localVarAxiosArgs, globalAxios, BASE_PATH, configuration);
        },
        /**
         * Deletes the resource and all its related data.
         * @summary Delete Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        async deleteResource(projId, envId, resourceId, options) {
            const localVarAxiosArgs = await localVarAxiosParamCreator.deleteResource(projId, envId, resourceId, options);
            return createRequestFunction(localVarAxiosArgs, globalAxios, BASE_PATH, configuration);
        },
        /**
         * Gets a single resource, if such resource exists.
         * @summary Get Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        async getResource(projId, envId, resourceId, options) {
            const localVarAxiosArgs = await localVarAxiosParamCreator.getResource(projId, envId, resourceId, options);
            return createRequestFunction(localVarAxiosArgs, globalAxios, BASE_PATH, configuration);
        },
        /**
         * Lists all the resources defined in your schema.
         * @summary List Resources
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {boolean} [includeBuiltIn] Whether to include or exclude built-in resources, default is False
         * @param {number} [page] Page number of the results to fetch, starting at 1.
         * @param {number} [perPage] The number of results per page (max 100).
         * @param {includeTotalCount} [includeTotalCount] Include the total count of resources in the response, default is False
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        async listResources(projId, envId, includeBuiltIn, page, perPage, includeTotalCount, options) {
            const localVarAxiosArgs = await localVarAxiosParamCreator.listResources(projId, envId, includeBuiltIn, page, perPage, includeTotalCount, options);
            return createRequestFunction(localVarAxiosArgs, globalAxios, BASE_PATH, configuration);
        },
        /**
         * Completely replaces the resource definition.  - If the resource key is changed, all role and permissions assignments for the the resource will be revoked. - If the resource key is unchanged, but some actions are removed or renamed from the resource definition, role and permissions assignments for these actions will be revoked.  TODO: we need to decide if we are auto-revoking, or if we are rejecting the PUT completely while there are permissions that can be affected.
         * @summary Replace Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {ResourceReplace} resourceReplace
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        async replaceResource(projId, envId, resourceId, resourceReplace, options) {
            const localVarAxiosArgs = await localVarAxiosParamCreator.replaceResource(projId, envId, resourceId, resourceReplace, options);
            return createRequestFunction(localVarAxiosArgs, globalAxios, BASE_PATH, configuration);
        },
        /**
         * Partially updates the resource definition. Fields that will be provided will be completely overwritten.
         * @summary Update Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {ResourceUpdate} resourceUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        async updateResource(projId, envId, resourceId, resourceUpdate, options) {
            const localVarAxiosArgs = await localVarAxiosParamCreator.updateResource(projId, envId, resourceId, resourceUpdate, options);
            return createRequestFunction(localVarAxiosArgs, globalAxios, BASE_PATH, configuration);
        },
    };
};
/**
 * ResourcesApi - factory interface
 * @export
 */
export const ResourcesApiFactory = function (configuration, basePath, axios) {
    const localVarFp = ResourcesApiFp(configuration);
    return {
        /**
         * Creates a new resource (a type of object you may protect with permissions).
         * @summary Create Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {ResourceCreate} resourceCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createResource(projId, envId, resourceCreate, options) {
            return localVarFp
                .createResource(projId, envId, resourceCreate, options)
                .then((request) => request(axios, basePath));
        },
        /**
         * Deletes the resource and all its related data.
         * @summary Delete Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteResource(projId, envId, resourceId, options) {
            return localVarFp
                .deleteResource(projId, envId, resourceId, options)
                .then((request) => request(axios, basePath));
        },
        /**
         * Gets a single resource, if such resource exists.
         * @summary Get Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getResource(projId, envId, resourceId, options) {
            return localVarFp
                .getResource(projId, envId, resourceId, options)
                .then((request) => request(axios, basePath));
        },
        /**
         * Lists all the resources defined in your schema.
         * @summary List Resources
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {boolean} [includeBuiltIn] Whether to include or exclude built-in resources, default is False
         * @param {number} [page] Page number of the results to fetch, starting at 1.
         * @param {number} [perPage] The number of results per page (max 100).
         * @param {includeTotalCount} [includeTotalCount] Include the total count of resources in the response, default is False
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        listResources(projId, envId, includeBuiltIn, page, perPage, includeTotalCount, options) {
            return localVarFp
                .listResources(projId, envId, includeBuiltIn, page, perPage, includeTotalCount, options)
                .then((request) => request(axios, basePath));
        },
        /**
         * Completely replaces the resource definition.  - If the resource key is changed, all role and permissions assignments for the the resource will be revoked. - If the resource key is unchanged, but some actions are removed or renamed from the resource definition, role and permissions assignments for these actions will be revoked.  TODO: we need to decide if we are auto-revoking, or if we are rejecting the PUT completely while there are permissions that can be affected.
         * @summary Replace Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {ResourceReplace} resourceReplace
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        replaceResource(projId, envId, resourceId, resourceReplace, options) {
            return localVarFp
                .replaceResource(projId, envId, resourceId, resourceReplace, options)
                .then((request) => request(axios, basePath));
        },
        /**
         * Partially updates the resource definition. Fields that will be provided will be completely overwritten.
         * @summary Update Resource
         * @param {string} projId Either the unique id of the project, or the URL-friendly key of the project (i.e: the \&quot;slug\&quot;).
         * @param {string} envId Either the unique id of the environment, or the URL-friendly key of the environment (i.e: the \&quot;slug\&quot;).
         * @param {string} resourceId Either the unique id of the resource, or the URL-friendly key of the resource (i.e: the \&quot;slug\&quot;).
         * @param {ResourceUpdate} resourceUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateResource(projId, envId, resourceId, resourceUpdate, options) {
            return localVarFp
                .updateResource(projId, envId, resourceId, resourceUpdate, options)
                .then((request) => request(axios, basePath));
        },
    };
};
/**
 * ResourcesApi - object-oriented interface
 * @export
 * @class ResourcesApi
 * @extends {BaseAPI}
 */
export class ResourcesApi extends BaseAPI {
    /**
     * Creates a new resource (a type of object you may protect with permissions).
     * @summary Create Resource
     * @param {ResourcesApiCreateResourceRequest} requestParameters Request parameters.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ResourcesApi
     */
    createResource(requestParameters, options) {
        return ResourcesApiFp(this.configuration)
            .createResource(requestParameters.projId, requestParameters.envId, requestParameters.resourceCreate, options)
            .then((request) => request(this.axios, this.basePath));
    }
    /**
     * Deletes the resource and all its related data.
     * @summary Delete Resource
     * @param {ResourcesApiDeleteResourceRequest} requestParameters Request parameters.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ResourcesApi
     */
    deleteResource(requestParameters, options) {
        return ResourcesApiFp(this.configuration)
            .deleteResource(requestParameters.projId, requestParameters.envId, requestParameters.resourceId, options)
            .then((request) => request(this.axios, this.basePath));
    }
    /**
     * Gets a single resource, if such resource exists.
     * @summary Get Resource
     * @param {ResourcesApiGetResourceRequest} requestParameters Request parameters.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ResourcesApi
     */
    getResource(requestParameters, options) {
        return ResourcesApiFp(this.configuration)
            .getResource(requestParameters.projId, requestParameters.envId, requestParameters.resourceId, options)
            .then((request) => request(this.axios, this.basePath));
    }
    /**
     * Lists all the resources defined in your schema.
     * @summary List Resources
     * @param {ResourcesApiListResourcesRequest} requestParameters Request parameters.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ResourcesApi
     */
    listResources(requestParameters, options) {
        return ResourcesApiFp(this.configuration)
            .listResources(requestParameters.projId, requestParameters.envId, requestParameters.includeBuiltIn, requestParameters.page, requestParameters.perPage, requestParameters.includeTotalCount, options)
            .then((request) => request(this.axios, this.basePath));
    }
    /**
     * Completely replaces the resource definition.  - If the resource key is changed, all role and permissions assignments for the the resource will be revoked. - If the resource key is unchanged, but some actions are removed or renamed from the resource definition, role and permissions assignments for these actions will be revoked.  TODO: we need to decide if we are auto-revoking, or if we are rejecting the PUT completely while there are permissions that can be affected.
     * @summary Replace Resource
     * @param {ResourcesApiReplaceResourceRequest} requestParameters Request parameters.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ResourcesApi
     */
    replaceResource(requestParameters, options) {
        return ResourcesApiFp(this.configuration)
            .replaceResource(requestParameters.projId, requestParameters.envId, requestParameters.resourceId, requestParameters.resourceReplace, options)
            .then((request) => request(this.axios, this.basePath));
    }
    /**
     * Partially updates the resource definition. Fields that will be provided will be completely overwritten.
     * @summary Update Resource
     * @param {ResourcesApiUpdateResourceRequest} requestParameters Request parameters.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ResourcesApi
     */
    updateResource(requestParameters, options) {
        return ResourcesApiFp(this.configuration)
            .updateResource(requestParameters.projId, requestParameters.envId, requestParameters.resourceId, requestParameters.resourceUpdate, options)
            .then((request) => request(this.axios, this.basePath));
    }
}
