// permitio/build/main/enforcement/enforcer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Enforcer = exports.PermitPDPStatusError = exports.PermitConnectionError = exports.PermitError = void 0;
const axios_1 = __importDefault(require("axios"));
const context_1 = require("../utils/context");
const http_logger_1 = require("../utils/http-logger");
const interfaces_1 = require("./interfaces");
const RESOURCE_DELIMITER = ':';
function isString(x) {
    return typeof x === 'string';
}
class PermitError extends Error {
    constructor(message) {
        super(message);
        this.name = 'PermitError';
    }
}
exports.PermitError = PermitError;
class PermitConnectionError extends PermitError {
    constructor(message) {
        super(message);
        this.name = 'PermitConnectionError';
    }
}
exports.PermitConnectionError = PermitConnectionError;
class PermitPDPStatusError extends PermitError {
    constructor(message) {
        super(message);
        this.name = 'PermitPDPStatusError';
    }
}
exports.PermitPDPStatusError = PermitPDPStatusError;
/**
 * The {@link Enforcer} class is responsible for performing permission checks against the PDP.
 * It implements the {@link IEnforcer} interface.
 */
class Enforcer {
    /**
     * Creates an instance of the Enforcer class.
     * @param config - The configuration object for the Permit SDK.
     * @param logger - The logger instance for logging.
     */
    constructor(config, logger) {
        var _a;
        this.config = config;
        this.logger = logger;
        const version = (_a = process.env.npm_package_version) !== null && _a !== void 0 ? _a : 'unknown';
        if (config.axiosInstance) {
            this.client = config.axiosInstance;
            this.client.defaults.baseURL = `${this.config.pdp}/`;
            this.client.defaults.headers.common['X-Permit-SDK-Version'] = `node:${version}`;
        }
        else {
            this.client = axios_1.default.create({
                baseURL: `${this.config.pdp}/`,
                headers: {
                    'X-Permit-SDK-Version': `node:${version}`,
                },
            });
        }
        this.logger = logger;
        http_logger_1.AxiosLoggingInterceptor.setupInterceptor(this.client, this.logger);
        this.contextStore = new context_1.ContextStore();
    }
    async getUserPermissions(user, tenants, resources, resource_types, config = {}) {
        return await this.getUserPermissionsWithExceptions(user, tenants, resources, resource_types, config).catch((err) => {
            const shouldThrow = config.throwOnError === undefined ? this.config.throwOnError : config.throwOnError;
            if (shouldThrow) {
                throw err;
            }
            else {
                this.logger.error(err);
                return {};
            }
        });
    }
    async getUserPermissionsWithExceptions(user, tenants, resources, resource_types, config = {}) {
        const checkTimeout = config.timeout || this.config.timeout;
        const input = {
            user: isString(user) ? { key: user } : user,
            tenants,
            resources,
            resource_types,
        };
        return await this.client
            .post('user-permissions', input, {
            headers: {
                Authorization: `Bearer ${this.config.token}`,
            },
            timeout: checkTimeout,
        })
            .then((response) => {
            if (response.status !== 200) {
                throw new PermitPDPStatusError(`Permit.getUserPermissions() got an unexpected status code: ${response.status}, please check your SDK init and make sure the PDP sidecar is configured correctly. \n\
            Read more about setting up the PDP at https://docs.permit.io`);
            }
            const permissions = ((0, interfaces_1.isOpaGetUserPermissionsResult)(response.data)
                ? response.data.result.permissions
                : response.data) || {};
            this.logger.info(`permit.getUserPermissions(${Enforcer.userRepr(input.user)}) = ${JSON.stringify(permissions)}`);
            return permissions;
        })
            .catch((error) => {
            var _a, _b;
            const errorMessage = `Error in permit.getUserPermissions(${Enforcer.userRepr(input.user)})`;
            if (axios_1.default.isAxiosError(error)) {
                const errorStatusCode = ((_a = error.response) === null || _a === void 0 ? void 0 : _a.status.toString()) || '';
                const errorDetails = ((_b = error === null || error === void 0 ? void 0 : error.response) === null || _b === void 0 ? void 0 : _b.data)
                    ? JSON.stringify(error.response.data)
                    : error.message;
                this.logger.error(`[${errorStatusCode}] ${errorMessage}, err: ${errorDetails}`);
            }
            else {
                this.logger.error(`${errorMessage}\n${error}`);
            }
            throw new PermitConnectionError(`Permit SDK got error: \n ${error.message} \n
          and cannot connect to the PDP, please check your configuration and make sure the
          PDP is running at ${this.config.pdp} and accepting requests. \n
          Read more about setting up the PDP at https://docs.permit.io`);
        });
    }
    async bulkCheck(checks, context = {}, // context provided specifically for this query
    config = {}) {
        return await this.bulkCheckWithExceptions(checks, context, config).catch((err) => {
            const shouldThrow = config.throwOnError === undefined ? this.config.throwOnError : config.throwOnError;
            if (shouldThrow) {
                throw err;
            }
            else {
                this.logger.error(err);
                return [];
            }
        });
    }
    buildCheckInput(user, action, resource, context = {}) {
        const normalizedUser = isString(user) ? { key: user } : user;
        const resourceObj = isString(resource) ? Enforcer.resourceFromString(resource) : resource;
        const normalizedResource = this.normalizeResource(resourceObj);
        const queryContext = this.contextStore.getDerivedContext(context);
        return {
            user: normalizedUser,
            action: action,
            resource: normalizedResource,
            context: queryContext,
        };
    }
    checkInputRepr(checkInput) {
        return `${Enforcer.userRepr(checkInput.user)}, ${checkInput.action}, ${Enforcer.resourceRepr(checkInput.resource)}`;
    }
    async bulkCheckWithExceptions(checks, context = {}, // context provided specifically for this query
    config = {}) {
        const checkTimeout = config.timeout || this.config.timeout;
        const inputs = [];
        checks.forEach((check) => {
            const input = this.buildCheckInput(check.user, check.action, check.resource, context);
            inputs.push(input);
        });
        return await this.client
            .post('allowed/bulk', inputs, {
            headers: {
                Authorization: `Bearer ${this.config.token}`,
            },
            timeout: checkTimeout,
        })
            .then((response) => {
            if (response.status !== 200) {
                throw new PermitPDPStatusError(`Permit.bulkCheck() got an unexpected status code: ${response.status}, please check your SDK init and make sure the PDP sidecar is configured correctly. \n\
            Read more about setting up the PDP at https://docs.permit.io`);
            }
            const decisions = (('allow' in response.data ? response.data.allow : response.data.result.allow) || []).map((decision) => decision.allow || false);
            this.logger.info(`permit.bulkCheck(${inputs.map((input) => this.checkInputRepr(input))}) = ${decisions}`);
            return decisions;
        })
            .catch((error) => {
            var _a, _b;
            const errorMessage = `Error in permit.bulkCheck(${inputs.map((input) => this.checkInputRepr(input))})`;
            if (axios_1.default.isAxiosError(error)) {
                const errorStatusCode = ((_a = error.response) === null || _a === void 0 ? void 0 : _a.status.toString()) || '';
                const errorDetails = ((_b = error === null || error === void 0 ? void 0 : error.response) === null || _b === void 0 ? void 0 : _b.data)
                    ? JSON.stringify(error.response.data)
                    : error.message;
                this.logger.error(`[${errorStatusCode}] ${errorMessage}, err: ${errorDetails}`);
            }
            else {
                this.logger.error(`${errorMessage}\n${error}`);
            }
            throw new PermitConnectionError(`Permit SDK got error: \n ${error.message} \n
          and cannot connect to the PDP, please check your configuration and make sure the
          PDP is running at ${this.config.pdp} and accepting requests. \n
          Read more about setting up the PDP at https://docs.permit.io`);
        });
    }
    async checkAllTenants(user, action, resource, context = {}, // default to empty context if not provided
    sdk = 'node') {
        try {
            const response = await this.client.post('/allowed/all-tenants', {
                headers: {
                    Authorization: `Bearer ${this.config.token}`,
                    'X-Permit-Sdk-Language': sdk,
                },
                params: {
                    user,
                    action,
                    resource,
                    context,
                },
            });
            return response.data.allowedTenants.map((item) => item.tenant);
        }
        catch (error) {
            this.logger.error('Error fetching all tenants:', error);
            throw error;
        }
    }
    async check(user, action, resource, context = {}, // context provided specifically for this query
    config = {}) {
        return await this.checkWithExceptions(user, action, resource, context, config).catch((err) => {
            const shouldThrow = config.throwOnError === undefined ? this.config.throwOnError : config.throwOnError;
            if (shouldThrow) {
                throw err;
            }
            else {
                this.logger.error(err);
                return false;
            }
        });
    }
    async checkWithExceptions(user, action, resource, context = {}, // context provided specifically for this query
    config = {}) {
        const input = this.buildCheckInput(user, action, resource, context);
        const checkTimeout = config.timeout || this.config.timeout;
        return await this.client
            .post('allowed', input, {
            headers: {
                Authorization: `Bearer ${this.config.token}`,
            },
            timeout: checkTimeout,
        })
            .then((response) => {
            if (response.status !== 200) {
                throw new PermitPDPStatusError(`Permit.check() got an unexpected status code: ${response.status}, please check your SDK init and make sure the PDP sidecar is configured correctly. \n\
            Read more about setting up the PDP at https://docs.permit.io`);
            }
            const decision = ('allow' in response.data ? response.data.allow : response.data.result.allow) || false;
            this.logger.info(`permit.check(${this.checkInputRepr(input)}) = ${decision}`);
            return decision;
        })
            .catch((error) => {
            var _a, _b;
            const errorMessage = `Error in permit.check(${this.checkInputRepr(input)})`;
            if (axios_1.default.isAxiosError(error)) {
                const errorStatusCode = ((_a = error.response) === null || _a === void 0 ? void 0 : _a.status.toString()) || '';
                const errorDetails = ((_b = error === null || error === void 0 ? void 0 : error.response) === null || _b === void 0 ? void 0 : _b.data)
                    ? JSON.stringify(error.response.data)
                    : error.message;
                this.logger.error(`[${errorStatusCode}] ${errorMessage}, err: ${errorDetails}`);
            }
            else {
                this.logger.error(`${errorMessage}\n${error}`);
            }
            throw new PermitConnectionError(`Permit SDK got error: \n ${error.message} \n
          and cannot connect to the PDP, please check your configuration and make sure the PDP is running at ${this.config.pdp} and accepting requests. \n
          Read more about setting up the PDP at https://docs.permit.io`);
        });
    }
    // TODO: remove this eventually, once we decide on finalized structure of AuthzQuery
    normalizeResource(resource) {
        const normalizedResource = Object.assign({}, resource);
        // if tenant is empty, we might auto-set the default tenant according to config
        if (!normalizedResource.tenant && this.config.multiTenancy.useDefaultTenantIfEmpty) {
            normalizedResource.tenant = this.config.multiTenancy.defaultTenant;
        }
        return normalizedResource;
    }
    static userRepr(user) {
        if (user.attributes || user.email) {
            return JSON.stringify(user);
        }
        return user.key;
    }
    static resourceRepr(resource) {
        var _a;
        if (resource.attributes && resource.attributes.length > 0) {
            return JSON.stringify(resource);
        }
        let resourceRepr = '';
        if (resource.tenant) {
            resourceRepr += `${resource.tenant}/`;
        }
        resourceRepr += `${resource.type}:${(_a = resource.key) !== null && _a !== void 0 ? _a : '*'}`;
        return resourceRepr;
    }
    static resourceFromString(resource) {
        const parts = resource.split(RESOURCE_DELIMITER);
        if (parts.length < 1 || parts.length > 2) {
            throw Error(`permit.check() got invalid resource string: '${resource}'`);
        }
        return {
            type: parts[0],
            key: parts.length > 1 ? parts[1] : undefined,
        };
    }
    getMethods() {
        return {
            check: this.check.bind(this),
            bulkCheck: this.bulkCheck.bind(this),
            getUserPermissions: this.getUserPermissions.bind(this),
            checkAllTenants: this.checkAllTenants.bind(this),
        };
    }
}
exports.Enforcer = Enforcer;
