// @permitio/cli/dist/commands/pdp/run.js
import React from 'react';
import { AuthProvider } from '../../components/AuthProvider.js';
import PDPRunComponent from '../../components/pdp/PDPRunComponent.js';
import { number, object, boolean, string } from 'zod';
import { option } from 'pastel';
export const description = 'Run a Permit PDP Docker container for local development';
export const options = object({
    opa: number()
        .optional()
        .describe(option({ description: 'Expose OPA port from the PDP' })),
    dryRun: boolean()
        .optional()
        .default(false)
        .describe(option({
        description: 'Print the Docker command without executing it',
        alias: 'd',
    })),
    apiKey: string()
        .optional()
        .describe(option({
        description: 'The API key for the Permit env, project or Workspace',
        alias: 'k',
    })),
});
export default function Run({ options: { opa, dryRun, apiKey } }) {
    return (React.createElement(AuthProvider, { permit_key: apiKey, scope: 'environment' },
        React.createElement(PDPRunComponent, { opa: opa, dryRun: dryRun })));
}
