// @permitio/cli/dist/commands/pdp/stats.js
import React from 'react';
import zod, { string } from 'zod';
import { option } from 'pastel';
import PDPStatComponent from '../../components/pdp/PDPStatComponent.js';
import { AuthProvider } from '../../components/AuthProvider.js';
export const options = zod.object({
    projectKey: zod
        .string()
        .optional()
        .describe(option({
        description: 'The project key',
        alias: 'p',
    })),
    environmentKey: zod
        .string()
        .optional()
        .describe(option({
        description: 'The environment key',
        alias: 'e',
    })),
    statsUrl: string()
        .optional()
        .describe(option({
        description: 'The URL of the PDP service. Default to the cloud PDP. (Optional)',
    })),
    apiKey: zod
        .string()
        .optional()
        .describe(option({
        description: 'The API key for the Permit env, project or Workspace (Optional)',
    })),
    top: zod.boolean().default(false).describe('Run stats in top mode'),
});
export default function Stats({ options }) {
    return (React.createElement(AuthProvider, { scope: 'environment', permit_key: options.apiKey },
        React.createElement(PDPStatComponent, { options: options })));
}
